src/Action/Smtp.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE ConstraintKinds        #-}
+{-# LANGUAGE DataKinds              #-}
+{-# LANGUAGE DeriveGeneric          #-}
+{-# LANGUAGE DefaultSignatures      #-}
+{-# LANGUAGE FlexibleInstances      #-}
+{-# LANGUAGE LambdaCase             #-}
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE OverloadedStrings      #-}
+{-# LANGUAGE ScopedTypeVariables    #-}
+{-# LANGUAGE TemplateHaskell        #-}
+
+module Action.Smtp
+    ( EmailMessage(..), msgSubject, msgBody, msgHtml
+    , HasSendMail(sendMailToAddress)
+    , SendMailError(..), ThrowSendMailError(..), MonadSendMailError
+    , sendMailToAddressIO
+    , sendMailToUser
+    , SendMailFlag(..)
+    -- , checkSendMail
+    ) where
+
+--import Control.Monad.Except (ExceptT, runExceptT)
+--import Control.Monad.Reader (ReaderT, runReaderT)
+import Thentos.Prelude hiding (logger)
+import Network.Mail.Mime (Address(Address), sendmailCustomCaptureOutput, simpleMail', renderMail')
+
+import qualified Data.ByteString as SB
+
+import Config
+import Types
+
+data SendMailError
+  = IOErrorRunningSendMail IOException
+  | NoEmailAddressForUser User
+  deriving (Eq)
+
+instance Show SendMailError where
+    show = \case
+        NoEmailAddressForUser user -> "No email address for user: " <> show (user ^. userLogin)
+        IOErrorRunningSendMail exn -> "IO error running sendmail: " ++ show exn
+
+class ThrowSendMailError err where
+    _SendMailError :: Prism' err SendMailError
+
+    throwSendMailError :: MonadError err m => SendMailError -> m any
+    throwSendMailError err = throwError $ _SendMailError # err
+
+instance ThrowSendMailError SendMailError where
+    _SendMailError = id
+
+type MonadSendMailError e m = (MonadError e m, ThrowSendMailError e)
+
+data EmailMessage = EmailMessage
+    { _msgSubject :: ST
+    , _msgBody    :: ST
+    , _msgHtml    :: Maybe ST
+    }
+    deriving (Eq, Ord, Show, Read, Generic)
+
+makeLenses ''EmailMessage
+
+data SendMailFlag = IgnoreMissingEmails
+    deriving (Eq, Ord, Show, Read, Generic)
+
+class (MonadSendMailError e m, MonadReaderConfig r m) => HasSendMail e r m where
+    sendMailToAddress :: Address -> EmailMessage -> m ()
+
+    default sendMailToAddress :: MonadIO m => Address -> EmailMessage -> m ()
+    sendMailToAddress = sendMailToAddressIO
```

Comment:
* The `Action` type is already in `MonadIO` so I'm using it. Again in the implementation the monad stack clearly has `IO` at the bottom `MonadIO` is here to help with that.
* If we were to remove the `MonadIO` from `Action` which I do not recommend but I'm not strongly opposed to that either, then we could have this one liner definition copied over `Action.Implementation`.
* I read this default implementation first as a hint and second as a shortcut if you have `MonadIO` anyway.